stdlib/path.gr

```diff
@@ -0,0 +1,729 @@
+  match (path) {
+    AbsoluteDir(x) => toUntyped(x),
+    AbsoluteFile(x) => toUntyped(x),
+    RelativeDir(x) => toUntyped(x),
+    RelativeFile(x) => toUntyped(x),
+  }
+}
+
+let toPath = (path: PathInfo) => {
+  match (path) {
+    (Abs(_), File, _) as p => AbsoluteFile(toTyped(p)),
+    (Abs(_), Dir, _) as p => AbsoluteDir(toTyped(p)),
+    (Rel(_), File, _) as p => RelativeFile(toTyped(p)),
+    (Rel(_), Dir, _) as p => RelativeDir(toTyped(p)),
+  }
+}
+
+// TODO expose these functions once module system added
+
+let absoluteFile = pathStr => {
+  let pathOpt = parseAbs(pathStr, File)
+  Option.map(toTyped, pathOpt): Option<TypedPath<Absolute, File>>
+}
+
+let absoluteDir = pathStr => {
+  let pathOpt = parseAbs(pathStr, Dir)
+  Option.map(toTyped, pathOpt): Option<TypedPath<Absolute, Directory>>
+}
+
+let parseRel = (pathStr, fileType) => {
+  let tokens = match (lexPath(pathStr)) {
+    [] => [Dot],
+    tokens => tokens,
+  }
+  let [first, ...rest] = tokens
+
+  if (fileType == File && !isFilePath(tokens)) {
+    None
+  } else {
+    let init = match (first) {
+      Dot => Some((Rel(0), fileType, [])),
+      Dotdot => Some((Rel(1), fileType, [])),
+      Text(str) => Some((Rel(0), fileType, [str])),
+      _ => None,
+    }
+    Option.map(init => List.reduce(parseNextToken, init, rest), init)
+  }
+}
+
+let relativeFile = pathStr => {
+  let pathOpt = parseRel(pathStr, File)
+  Option.map(toTyped, pathOpt): Option<TypedPath<Relative, File>>
+}
+
+let relativeDir = pathStr => {
+  // parseRel(pathStr, Dir)//: Option<TypedPath<Relative, Directory>>
+  let pathOpt = parseRel(pathStr, Dir)
+  Option.map(toTyped, pathOpt): Option<TypedPath<Relative, Directory>>
+}
+
+/**
+ * Parses a path string into a path wrapped with the path's type. If the path
+ * type is known in advance, the specialized parsing functions (`relativeDir`,
+ * `absoluteFile`, etc) should be used instead.
+ * 
+ * @param pathStr: The path string to parse
+ * @returns `Some(wrappedPath)` if the path is successfully parsed or `None` otherwise
+ * 
+ * @example fromString("/bin/") // an absolute Path referencing the directory /bin/
+ * @example fromString("file.txt") // a relative Path referencing the file ./file.txt
+ * @example fromString(".") // a relative Path referencing the current directory
+ * 
+ * @since v0.5.5
+ */
+export let fromString = pathStr => {
+  let tokens = match (lexPath(pathStr)) {
+    // will cause empty strings to get parsed as relative directory '.'
+    [] => [Dot],
+    tokens => tokens,
+  }
+  let isAbs = match (tokens) {
+    [Slash | DriveTok(_), ..._] => true,
+    _ => false,
+  }
+  let isDir = !isFilePath(tokens)
+
+  match ((isAbs, isDir)) {
+    (true, true) => AbsoluteDir(Option.unwrap(absoluteDir(pathStr))),
+    (true, false) => AbsoluteFile(Option.unwrap(absoluteFile(pathStr))),
+    (false, true) => RelativeDir(Option.unwrap(relativeDir(pathStr))),
+    (false, false) => RelativeFile(Option.unwrap(relativeFile(pathStr))),
```

Comment:
Should these `unwrap`s be `expect` so it can throw `Impossible:` messages if encountered?